test/hotspot/jtreg/testlibrary_tests/ir_framework/tests/TestSetupTests.java

```diff
@@ -0,0 +1,346 @@
+            }
+        }
+    }
+
+    // ---------- Setup "linked" random values ---------------
+    @Setup
+    static Object[] setupLinkedII() {
+        int r = RANDOM.nextInt();
+        return new Object[]{ r, r + 42};
+    }
+
+    @Test
+    @Arguments(setup = "setupLinkedII")
+    static int testSetupLinkedII(int a, int b) {
+        return b - a;
+    }
+
+    @Check(test = "testSetupLinkedII")
+    static void checkSetupLinkedII(int res) {
+        if (res != 42) { throw new RuntimeException("wrong result " + res); }
+    }
+}
+
+class TestSetupTestsWithFields {
+    int iFld1, iFld2, iFld3;
+
+    @Setup
+    Object[] setupTest1(SetupInfo info) {
+        iFld1 = info.invocationCounter() + 1;
+        iFld2 = info.invocationCounter() + 2;
+        iFld3 = info.invocationCounter() + 3;
+        return new Object[]{info.invocationCounter()}; // -> argument x in test
+    }
+
+    @Test
+    @Arguments(setup = "setupTest1")
+    int test1(int x) {
+        if (iFld1 != x + 1) { throw new RuntimeException("iFld1 wrong value: " + iFld1 + " != " + (x + 1)); }
+        if (iFld2 != x + 2) { throw new RuntimeException("iFld2 wrong value: " + iFld2 + " != " + (x + 2)); }
+        if (iFld3 != x + 3) { throw new RuntimeException("iFld3 wrong value: " + iFld3 + " != " + (x + 3)); }
+        iFld1++;
+        iFld2++;
+        iFld3++;
+        return x + 5; // -> argument y in check
+    }
+
+    @Check(test = "test1")
+    void checkTest1(int y) {
+        if (iFld1 != y - 3) { throw new RuntimeException("iFld1 wrong value: " + iFld1 + " != " + (y - 3)); }
+        if (iFld2 != y - 2) { throw new RuntimeException("iFld2 wrong value: " + iFld2 + " != " + (y - 2)); }
+        if (iFld3 != y - 1) { throw new RuntimeException("iFld3 wrong value: " + iFld3 + " != " + (y - 1)); }
+    }
+}
+
+class TestSetupTestsSetupInfo {
+    static int lastCnt = -1;
+
+    @Setup
+    Object[] setupTest1(SetupInfo info) {
+        int cnt = info.invocationCounter();
+        // Check that we increment every time
+        if (cnt - 1 != lastCnt) {
+            throw new RuntimeException("SetupInfo invocationCounter does not increment correctly: " +
+                                       cnt + ", vs last: " + lastCnt);
+        }
+        lastCnt = cnt;
+        return new Object[]{1, 2};
+    }
+
+    @Test
+    @Arguments(setup = "setupTest1")
+    void test1(int a, int b) {}
+}
+
+class TestSetupTestsWithExpectedExceptions {
+    // ----------------- wrong number of arguments ------------------
+    @Setup
+    public Object[] setupTooManyArgs() {
+      return new Object[]{1, 2, 3};
+    }
+
+    @Test
+    @Arguments(setup = "setupTooManyArgs")
+    public void testTooManyArgs(int a) {}
+
+    @Setup
+    public Object[] setupTooFewArgs() {
+      return new Object[]{1, 2};
+    }
+
+    @Test
+    @Arguments(setup = "setupTooFewArgs")
+    public void testTooFewArgs(int a, int b, int c) {}
```

Comment:
We should also test:
- Wrongly returning an empty Object array where we expect at least one argument
- Having a `@Setup` method even though the `@Test` method takes no arguments.

But I might have missed that you already provided such tests .

Reply:
will do